File: frontend/src/utils/vault.ts
import { VaultItem } from '../types/download';

export interface MatchedVaultResult {
  host: string;
  user: string;
  pass: string;
  username: string;
  password: string;
  userAgent: string;
  speedLimit: number; // in bytes/sec
  threadCount: number;
}

export const matchVaultCredentials = (
  targetUrl: string,
  vaultList: VaultItem[] = []
): MatchedVaultResult | null => {
  if (!targetUrl || !vaultList || !vaultList.length) return null;
  try {
    const parsed = new URL(targetUrl);
    const urlHost = parsed.host.toLowerCase(); // e.g. "localhost:3000"
    const urlHostname = parsed.hostname.toLowerCase(); // e.g. "localhost"
    const urlSchemeHost = `${parsed.protocol}//${parsed.host}`.toLowerCase(); // e.g. "http://localhost:3000"

    for (const item of vaultList) {
      if (!item || !item.host) continue;
      let vHost = item.host.trim().toLowerCase().replace(/\/+$/, '');
      if (!vHost) continue;

      let isMatch = false;

      // 1. Exact string match against scheme+host, host, or hostname
      if (vHost === urlSchemeHost || vHost === urlHost || vHost === urlHostname) {
        isMatch = true;
      }
      // 2. If vHost contains protocol (e.g. "http://localhost:3000" or "https://example.com")
      else if (vHost.startsWith('http://') || vHost.startsWith('https://')) {
        try {
          const vParsed = new URL(vHost);
          if (vParsed.hostname.toLowerCase() === urlHostname) {
            if (!vParsed.port || vParsed.port === parsed.port) {
              isMatch = true;
            }
          }
        } catch {}
      }
      // 3. Domain or wildcard match
      else {
        const cleanWildcard = vHost.replace(/^\*\./, '');
        const hostWithoutPort = cleanWildcard.split(':')[0];
        if (vHost === urlHost || urlHostname === hostWithoutPort || urlHostname.endsWith('.' + hostWithoutPort)) {
          isMatch = true;
        }
      }

      if (isMatch) {
        const u = item.user || (item as any).username || '';
        const p = item.pass || (item as any).password || '';
        return {
          host: item.host,
          user: u,
          pass: p,
          username: u,
          password: p,
          userAgent: item.userAgent || (item as any).user_agent || '',
          speedLimit: item.speedLimit || (item as any).speed_limit || 0,
          threadCount: item.threadCount || (item as any).thread_count || 0,
        };
      }
    }
  } catch {}
  return null;
};

File: src-wails3/downloader/vault.go
package downloader

import (
	"encoding/json"
	"net/url"
	"strings"
	"ThunderDM/src-wails3/storage"
)

type VaultItem struct {
	Host        string `json:"host"`
	User        string `json:"user,omitempty"`
	Pass        string `json:"pass,omitempty"`
	SpeedLimit  int64  `json:"speedLimit,omitempty"`
	ThreadCount int    `json:"threadCount,omitempty"`
	UserAgent   string `json:"userAgent,omitempty"`
}

// MatchVaultItem checks ~/.thunderdm/download_engine.json or in-memory config for matching per-host settings for the given URL.
func MatchVaultItem(rawURL string) *VaultItem {
	if rawURL == "" {
		return nil
	}

	parsed, err := url.Parse(rawURL)
	if err != nil {
		return nil
	}

	var items []VaultItem
	cachedCfg := GetEngineConfig()
	if len(cachedCfg.VaultItems) > 0 {
		items = cachedCfg.VaultItems
	} else {
		if val, err := storage.GetKV("download_engine"); err == nil && len(val) > 0 {
			var root map[string]interface{}
			if json.Unmarshal([]byte(val), &root) == nil {
				if vaultRaw, ok := root["vaultItems"]; ok && vaultRaw != nil {
					if rawBytes, mErr := json.Marshal(vaultRaw); mErr == nil {
						_ = json.Unmarshal(rawBytes, &items)
					}
				}
			}
		}
	}

	if len(items) == 0 {
		return nil
	}

	urlHost := strings.ToLower(parsed.Host)                               // e.g. "localhost:3000"
	urlHostname := strings.ToLower(parsed.Hostname())                     // e.g. "localhost"
	urlSchemeHost := strings.ToLower(parsed.Scheme + "://" + parsed.Host) // e.g. "http://localhost:3000"

	for _, item := range items {
		vHost := strings.TrimSpace(strings.ToLower(item.Host))
		if vHost == "" {
			continue
		}
		// Strip trailing slash
		vHost = strings.TrimRight(vHost, "/")

		// Case 1: Exact scheme+host match (e.g. "http://localhost:3000")
		if vHost == urlSchemeHost {
			itemCopy := item
			return &itemCopy
		}
		// Case 2: Exact host:port match (e.g. "localhost:3000")
		if vHost == urlHost {
			itemCopy := item
			return &itemCopy
		}
		// Case 3: Exact hostname match (e.g. "localhost" or "drive.google.com")
		if vHost == urlHostname {
			itemCopy := item
			return &itemCopy
		}
		// Case 4: vHost has scheme with hostname (e.g. "http://localhost")
		if strings.HasPrefix(vHost, "http://") || strings.HasPrefix(vHost, "https://") {
			if vParsed, pErr := url.Parse(vHost); pErr == nil {
				if strings.EqualFold(vParsed.Hostname(), urlHostname) {
					if vParsed.Port() == "" || vParsed.Port() == parsed.Port() {
						itemCopy := item
						return &itemCopy
					}
				}
			}
		}
		// Case 5: Domain wildcard or domain match (e.g. "*.example.com" or "example.com")
		cleanVHost := strings.TrimPrefix(vHost, "*.")
		if strings.HasSuffix(urlHostname, cleanVHost) {
			itemCopy := item
			return &itemCopy
		}
	}

	return nil
}

// MatchVaultCredentials checks ~/.thunderdm/download_engine.json or in-memory config for matching credentials for the given URL.
func MatchVaultCredentials(rawURL string) (string, string) {
	item := MatchVaultItem(rawURL)
	if item != nil {
		return item.User, item.Pass
	}
	return "", ""
}
